libraries/lib-ffmpeg-support/wrappers/AVCodecContextWrapper.cpp

```diff
@@ -0,0 +1,180 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  AVCodecContextWrapper.cpp
+
+  Dmitry Vedenko
+
+**********************************************************************/
+
+#include "AVCodecContextWrapper.h"
+
+#include <cstring>
+
+#include "FFmpegFunctions.h"
+#include "AVCodecWrapper.h"
+
+AVCodecContextWrapper::AVCodecContextWrapper(
+   const FFmpegFunctions* ffmpeg, std::unique_ptr<AVCodecWrapper> codec) noexcept
+    : mFFmpeg(ffmpeg)
+    , mAVCodec(std::move(codec))
+    , mIsOwned(true)
+{
+   if (mFFmpeg == nullptr)
+      return;
+
+   mAVCodecContext =
+      mFFmpeg->avcodec_alloc_context3(mAVCodec->GetWrappedValue());
+}
+
+AVCodecContextWrapper::AVCodecContextWrapper(
+   const FFmpegFunctions* ffmpeg, AVCodecContext* wrapped) noexcept
+    : mFFmpeg(ffmpeg)
+    , mAVCodecContext(wrapped)
+    , mIsOwned(false)
+{
+
+}
+
+AVCodecContext* AVCodecContextWrapper::GetWrappedValue() noexcept
+{
+    return mAVCodecContext;
+}
+
+const AVCodecContext* AVCodecContextWrapper::GetWrappedValue() const noexcept
+{
+   return mAVCodecContext;
+}
+
+AVCodecContextWrapper::~AVCodecContextWrapper()
+{
+   if (mIsOwned && mFFmpeg != nullptr && mAVCodecContext != nullptr)
+   {
+      // avcodec_free_context, complementary to avcodec_alloc_context3, is
+      // not necessarily loaded
+      if (mFFmpeg->avcodec_free_context != nullptr)
+      {
+         mFFmpeg->avcodec_free_context(&mAVCodecContext);
+      }
+      else
+      {
+         // Its not clear how to avoid the leak here, but let's close
+         // the codec at least
+         if (mFFmpeg->avcodec_is_open(mAVCodecContext))
+            mFFmpeg->avcodec_close(mAVCodecContext);
+      }
+   }
+}
+
+std::vector<uint8_t>
+AVCodecContextWrapper::DecodeAudioPacket(const AVPacketWrapper* packet)
+{
+   if (mFFmpeg == nullptr)
+      return {};
+
+   auto frame = mFFmpeg->CreateAVFrameWrapper();
+   std::vector<uint8_t> data;
+
+   if (mFFmpeg->avcodec_decode_audio4)
+   {
+      std::unique_ptr<AVPacketWrapper> packetCopy =
+         packet ? packet->Clone() : mFFmpeg->CreateAVPacketWrapper();
+
+      /*
+       "Flushing is done by calling this function [avcodec_decode_audio4]
+       with packets with avpkt->data set to NULL and avpkt->size set to 0
+       until it stops returning samples."
+       (That implies, not necessarily just one loop pass to flush)
+       */
+      bool flushing = packet
+         ? (packetCopy->GetSize() == 0 && packetCopy->GetData() == nullptr)
+         : true;
+      if (!flushing && packetCopy->GetData() == nullptr)
+         return {};
+
+      const int channels = GetChannels();
+
+      if (flushing)
+         GetChannels();
```

Comment:
These two new lines have no side effect and just repeat the GetChannels call.  Did you intend something else?

Reply:
Debugging leftovers, fixed. Conditional breakpoints are too slow for practical use :(